cache/lib/cache/s3.ex

```diff
@@ -46,4 +46,101 @@ defmodule Cache.S3 do
       {:error, _reason} -> false
     end
   end
+
+  @doc """
+  Uploads an artifact to S3.
+
+  Returns :ok on success or {:error, reason} on failure.
+  If the local file does not exist, returns :ok (the file may have been evicted).
+  """
+  def upload(account_handle, project_handle, artifact_id) do
+    key = Cache.Disk.cas_key(account_handle, project_handle, artifact_id)
+    local_path = Cache.Disk.artifact_path(key)
+
+    Logger.info("Starting S3 upload for artifact: #{key}")
+
+    case upload_file(key, local_path) do
+      :ok ->
+        Logger.info("Successfully uploaded artifact to S3: #{key}")
+        :ok
+
+      {:error, reason} ->
+        Appsignal.send_error(%RuntimeError{message: "Failed to upload artifact to S3"}, %{
+          key: key,
+          reason: reason
+        })
+
+        {:error, reason}
+    end
+  end
+
+  defp upload_file(key, local_path) do
+    if File.exists?(local_path) do
+      bucket = Application.get_env(:cache, :s3)[:bucket]
+
+      case local_path
+           |> ExAws.S3.Upload.stream_file()
+           |> ExAws.S3.upload(bucket, key)
+           |> ExAws.request() do
+        {:ok, _response} ->
+          :ok
+
+        {:error, reason} ->
+          {:error, reason}
+      end
+    else
+      Logger.error("Local file not found for S3 upload: #{local_path}")
+      :ok
+    end
+  end
+
+  @doc """
+  Downloads an artifact from S3 to local disk.
+
+  Returns :ok on success or {:error, reason} on failure.
+  If the artifact does not exist in S3, returns :ok (nothing to download).
+  """
+  def download(account_handle, project_handle, artifact_id) do
+    key = Cache.Disk.cas_key(account_handle, project_handle, artifact_id)
+    Logger.info("Starting S3 download for artifact: #{key}")
+
+    if exists?(key) do
+      local_path = Cache.Disk.artifact_path(key)
+
+      case download_file(key, local_path) do
+        :ok ->
+          {:ok, %{size: size}} = Cache.Disk.stat(account_handle, project_handle, artifact_id)
+
+          :telemetry.execute([:cache, :cas, :download, :s3_hit], %{size: size}, %{
```

Comment:
should we track misses, too, or would we count misses as number of download requests - hits?